include/dlaf/communication/kernels/all_reduce.h

```diff
@@ -0,0 +1,151 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2021, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <mpi.h>
+
+#include "dlaf/common/bag.h"
+#include "dlaf/common/callable_object.h"
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/communication/executor.h"
+#include "dlaf/communication/message.h"
+#include "dlaf/communication/rdma.h"
+#include "dlaf/executors.h"
+#include "dlaf/matrix/tile.h"
+
+namespace dlaf {
+namespace comm {
+
+namespace internal {
+
+template <class T>
+auto allReduce(common::PromiseGuard<comm::Communicator> pcomm, MPI_Op reduce_op,
+               common::internal::Bag<const T> bag_in, common::internal::Bag<T> bag_out,
+               matrix::Tile<const T, Device::CPU> const&, MPI_Request* req) {
+  auto& comm = pcomm.ref();
+  auto msg_in = comm::make_message(hpx::get<1>(bag_in));
+  auto msg_out = comm::make_message(hpx::get<1>(bag_out));
+
+  DLAF_MPI_CALL(MPI_Iallreduce(msg_in.data(), msg_out.data(), msg_in.count(), msg_in.mpi_type(),
+                               reduce_op, comm, req));
+
+  return bag_out;
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(allReduce);
+
+template <class T>
+auto allReduceInPlace(common::PromiseGuard<comm::Communicator> pcomm, MPI_Op reduce_op,
+                      common::internal::Bag<T> bag, MPI_Request* req) {
+  auto& comm = pcomm.ref();
+  auto msg = comm::make_message(hpx::get<1>(bag));
+
+  DLAF_MPI_CALL(
+      MPI_Iallreduce(MPI_IN_PLACE, msg.data(), msg.count(), msg.mpi_type(), reduce_op, comm, req));
+
+  return bag;
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(allReduceInPlace);
+}
+
+template <class T>
+void scheduleAllReduce(const comm::Executor& ex,
+                       hpx::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
+                       hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile_in,
+                       hpx::future<matrix::Tile<T, Device::CPU>> tile_out) {
+  // Note:
+  //
+  //         +---------------------------------+
+  //         |                                 |
+  // TILE_I -+-> makeContiguous -----> BAG_I --+--> mpi_call --> BAG_O --+
+  //                                           |                         |
+  // TILE_O ---> makeContiguous --+--> BAG_O --+                         |
+  //                              |                                      |
+  //                              +----------------> TILE_O -------------+-> copyBack
```

Comment:
:+1:

More of this, I like it!

Reply:
😄 Just for reference, I used this https://asciiflow.com